File: lib/data/manual-review-queue.ts
export type ManualReviewQueueKind =
  | "institution_identity"
  | "programme_offering_identity"
  | "applicant_pathway"
  | "source_extraction"
  | "data_completeness"
  | "general"

export type ManualReviewReason = {
  reason: string
  queue: ManualReviewQueueKind
}

export type ManualReviewRecord = {
  needsReview?: boolean
  reviewReasons?: readonly string[]
}

export function classifyManualReviewReason(reason: string): ManualReviewReason {
  if (/alias|identity|campus|institution/i.test(reason)) {
    return { reason, queue: "institution_identity" }
  }
  if (/programme.*(name|title)|title_leak|requirement_fragment/i.test(reason)) {
    return { reason, queue: "programme_offering_identity" }
  }
  if (/equivalent|pathway|route|accepts/i.test(reason)) {
    return { reason, queue: "applicant_pathway" }
  }
  if (/pdf|extraction|dotted_filler|parsed/i.test(reason)) {
    return { reason, queue: "source_extraction" }
  }
  if (/missing|unknown|confidence/i.test(reason)) {
    return { reason, queue: "data_completeness" }
  }
  return { reason, queue: "general" }
}

export function mergeReviewReasons(values: Array<string | undefined>) {
  return [
    ...new Set(
      values
        .flatMap((value) => value?.split(/[;,|]/) ?? [])
        .map((value) => value.trim())
        .filter(Boolean)
    ),
  ]
}

export function needsManualReview(reasons: readonly string[]) {
  return reasons.length > 0
}

export function summarizeManualReviewQueues(
  records: readonly ManualReviewRecord[]
) {
  const queues: Record<ManualReviewQueueKind, number> = {
    institution_identity: 0,
    programme_offering_identity: 0,
    applicant_pathway: 0,
    source_extraction: 0,
    data_completeness: 0,
    general: 0,
  }

  for (const record of records) {
    for (const reason of record.reviewReasons ?? []) {
      queues[classifyManualReviewReason(reason).queue] += 1
    }
  }

  return {
    needsReview: records.filter((record) => record.needsReview).length,
    queues,
  }
}

File: analysis/src/kozi_analysis/manual_review_queue.py
from __future__ import annotations

import re
from collections import Counter
from typing import Any, Literal

ManualReviewQueueKind = Literal[
    "institution_identity",
    "programme_offering_identity",
    "applicant_pathway",
    "source_extraction",
    "data_completeness",
    "general",
]


def classify_manual_review_queue(reason: str) -> ManualReviewQueueKind:
    if re.search(r"alias|identity|campus|institution", reason, re.I):
        return "institution_identity"
    if re.search(
        r"programme.*(name|title)|title_leak|requirement_fragment", reason, re.I
    ):
        return "programme_offering_identity"
    if re.search(r"equivalent|pathway|route|accepts", reason, re.I):
        return "applicant_pathway"
    if re.search(r"pdf|extraction|dotted_filler|parsed", reason, re.I):
        return "source_extraction"
    if re.search(r"missing|unknown|confidence", reason, re.I):
        return "data_completeness"
    return "general"


def manual_review_queue_distribution(records: list[dict[str, Any]]) -> Counter[str]:
    counter: Counter[str] = Counter()
    for record in records:
        reasons = record.get("reviewReasons")
        if isinstance(reasons, list):
            counter.update(
                classify_manual_review_queue(str(reason)) for reason in reasons
            )
    return counter
